### PR TITLE
Warn when a resource with content auto expiry don't have a explicit RetentionPeriod

### PR DESCRIPTION
**Pull request by mmouniro**
*Issue #, if available:* https://github.com/aws-cloudformation/cfn-python-lint/issues/19

The behavior for data retention is different across AWS Services.
If no retention period is specified the default for some services is to delete the data after a period of time.

This check warn the template author to explicitly set the retention period to avoid unexpected data losses.

*Description of changes:*

Created a new *warn* rule validating 7 resources for which the retention period should is explicitly set. 
These resources were chosen for validation because their default behavior is to **automatically** delete the content after a period of time. Such an example is SQS where messages in the queue will be deleted by default after 4 days

Resources that retain the data forever when the retention is not set are not part of the validation.

*Tests*

I've executed the following commands

* `pip3 install -e .`
* `tox`
* `cfn-lint --update-documentation`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/fixtures/templates/bad/resources/sqs/retention_period.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
```
Comment (PatMyron) [inline test/fixtures/templates/bad/resources/sqs/retention_period.yml:1]: Extension from `.yml` to [`.yaml`](https://travis-ci.com/github/aws-cloudformation/cfn-python-lint/jobs/367200676#L300) for this test:

https://github.com/aws-cloudformation/cfn-python-lint/blob/bc65cb3d160486cbc664664f752ed9399a4234fd/test/unit/rules/resources/test_retentionperiod.py#L27

File: test/fixtures/templates/good/resources/sqs/retention_period.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+---
+AWSTemplateFormatVersion: "2010-09-09"
+Resources: 
+  MyQueue: 
+    Type: AWS::SQS::Queue
+    Properties: 
+      QueueName: "SampleQueue"
+      MessageRetentionPeriod: 5
```
Comment (PatMyron) [inline test/fixtures/templates/good/resources/sqs/retention_period.yaml:8]: Thanks for adding a valid template as well, [looks like this picked up a false positive](https://travis-ci.com/github/aws-cloudformation/cfn-python-lint/jobs/367200676#L354)

Comment (PatMyron) [inline test/fixtures/templates/good/resources/sqs/retention_period.yaml:8]:
```suggestion
      MessageRetentionPeriod: 60
```

Should fix this unrelated error while we're at it so the good template has no violations:
`E3034 Value has to be between 60 and 1209600 at Resources/MyQueue/Properties/MessageRetentionPeriod`

**Comment by PatMyron** [comment]
looks like this is picking up some valid violations in some of our other test fixtures as well:
https://github.com/aws-cloudformation/cfn-python-lint/blob/961ec8f026c6b7774e83e6ca73ca5336f5c1c03a/test/fixtures/templates/quickstart/nist_application.yaml#L1002-L1024
will need to acknowledge those as valid violations to get the Travis tests passing again:
[`test/fixtures/results/quickstart/nist_application.json`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/test/fixtures/results/quickstart/nist_application.json)
[`test/fixtures/results/quickstart/non_strict/nist_application.json`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/test/fixtures/results/quickstart/non_strict/nist_application.json)
(might be some other spots as well. We can help with this if needed since it's a bit unwieldy at first)

**Comment by PatMyron** [comment]
Other `retention` properties from searching the **[CloudFormation Resource Specification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html) to consider:**

AWS::ElastiCache::CacheCluster.SnapshotRetentionLimit
AWS::ElastiCache::ReplicationGroup.SnapshotRetentionLimit
AWS::FSx::FileSystem.LustreConfiguration.AutomaticBackupRetentionDays
AWS::FSx::FileSystem.WindowsConfiguration.AutomaticBackupRetentionDays
AWS::Glue::Table.TableInput.Retention
AWS::IoTAnalytics::Channel.RetentionPeriod
AWS::IoTAnalytics::Dataset.RetentionPeriod
AWS::IoTAnalytics::Datastore.RetentionPeriod
AWS::Neptune::DBCluster.BackupRetentionPeriod
AWS::S3::Bucket.DefaultRetention
AWS::Logs::LogGroup.RetentionDays

**Comment by mmouniro** [comment]
> Other `retention` properties from searching the **[CloudFormation Resource Specification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html) to consider:**
> 
> [`AWS::ElastiCache::CacheCluster.SnapshotRetentionLimit`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticache-cache-cluster.html#cfn-elasticache-cachecluster-snapshotretentionlimit)
> [`AWS::ElastiCache::ReplicationGroup.SnapshotRetentionLimit`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticache-replicationgroup.html#cfn-elasticache-replicationgroup-snapshotretentionlimit)
> [`AWS::FSx::FileSystem.LustreConfiguration.AutomaticBackupRetentionDays`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-lustreconfiguration.html#cfn-fsx-filesystem-lustreconfiguration-automaticbackupretentiondays)
> [`AWS::FSx::FileSystem.WindowsConfiguration.AutomaticBackupRetentionDays`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-windowsconfiguration.html#cfn-fsx-filesystem-windowsconfiguration-automaticbackupretentiondays)
> [`AWS::Glue::Table.TableInput.Retention`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-glue-table-tableinput.html#cfn-glue-table-tableinput-retention)
> [`AWS::IoTAnalytics::Channel.RetentionPeriod`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iotanalytics-channel-retentionperiod.html)
> [`AWS::IoTAnalytics::Dataset.RetentionPeriod`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iotanalytics-dataset-retentionperiod.html)
> [`AWS::IoTAnalytics::Datastore.RetentionPeriod`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iotanalytics-datastore-retentionperiod.html)
> [`AWS::Logs::LogGroup.RetentionDays`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-logs-loggroup.html#cfn-cwl-loggroup-retentionindays)
> [`AWS::Neptune::DBCluster.BackupRetentionPeriod`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-neptune-dbcluster.html#cfn-neptune-dbcluster-backupretentionperiod)
> [`AWS::S3::Bucket.DefaultRetention`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-defaultretention.html)


The reason I removed those is that when the retention is not set those serves will **keep** the data leading to no data looses. 
In that case it doesn't feel like we should warn the author on the lack of retention period. 

Willing to change if you feel that it should always warn

**Comment by mmouniro** [comment]
> looks like this is picking up some valid violations in some of our other test fixtures as well:
> https://github.com/aws-cloudformation/cfn-python-lint/blob/961ec8f026c6b7774e83e6ca73ca5336f5c1c03a/test/fixtures/templates/quickstart/nist_application.yaml#L1002-L1024
> 
> 
> will need to acknowledge those as valid violations to get the Travis tests passing again:
> [`test/fixtures/results/quickstart/nist_application.json`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/test/fixtures/results/quickstart/nist_application.json)
> [`test/fixtures/results/quickstart/non_strict/nist_application.json`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/test/fixtures/results/quickstart/non_strict/nist_application.json)
> (might be some other spots as well. We can help with this if needed since it's a bit unwieldy at first)

What is the preferred approach here ? Fix the template or add an exceptions ?

**Comment by kddejong** [comment]
Closed in favor of #2054

**Pull request closed without merging**